examples/nlp/gpt/conf/gpt_spin.yaml

```diff
@@ -0,0 +1,182 @@
+defaults:
+  - optional tp_overlap@model.ub_tp_comm_overlap_cfg:
+
+trainer:
+  num_nodes: 8
+  devices: 8
+  accelerator: gpu
+  precision: bf16-mixed
+  
+  # spin specific args
+  spin:
+    max_iterations: 3
+    max_epochs: 1  # use this to control the number of episodes played (T in the original paper)
+    max_steps: -1
+    val_check_interval: 0.1
+    save_interval: 100
+
+    # how many GBS we loop over
+    limit_val_batches: 1.0
+    gradient_clip_val: 1.0
+
+  # do not change these
+  logger: False # logger provided by exp_manager
+  enable_checkpointing: False
+  use_distributed_sampler: False
+  max_time: '00:03:45:00'
+  max_epochs: ${.spin.max_epochs}
+  max_steps: ${.spin.max_steps}
+
+exp_manager:
+  explicit_log_dir: /results
+  exp_dir: null
+  name: megatron_gpt
+  max_time_per_run: ${trainer.max_time}
+  create_wandb_logger: False
+  wandb_logger_kwargs:
+    project: nemo_aligner_dpo
+    name: rlhf_gpt3_dpo
+  resume_from_checkpoint: null # The path to a checkpoint file to continue the training, restores the whole state including the epoch, step, LR schedulers, apex, etc.
+  resume_if_exists: True
+  resume_ignore_no_checkpoint: True
+  create_checkpoint_callback: True
+  checkpoint_callback_params:
+    monitor: val_loss
+    save_top_k: 3
+    mode: min
+    always_save_nemo: False # saves nemo file during validation, not implemented for model parallel
+    save_nemo_on_train_end: True # not recommended when training large models on clusters with short time limits
+    filename: 'megatron_gpt--{${.monitor}:.3f}-{step}-{consumed_samples}-{iteration}-{epoch}'
+    model_parallel_size: ${multiply:${model.tensor_model_parallel_size}, ${model.pipeline_model_parallel_size}}
+
+pretrained_checkpoint:
+  restore_from_path: null
+
+model:  
+  mcore_gpt: True
+  micro_batch_size: 1
+  global_batch_size: 64
+  megatron_amp_O2: True
+
+  spin:
+    log_prob_forward_micro_batch_size: 1
+    ref_policy_kl_penalty: 0.2
+    offload_adam_states: True
+
+    # params for generation
+    sampling_params:
+      use_greedy: False
+      temperature: 1.0
+      top_k: 0
+      top_p: 1.0
+      repetition_penalty: 1.0
+      add_BOS: False
+      all_probs: False
+      compute_logprob: False
+      # will be used in NeMo version > 1.20.0
+      # keeping it for now
+      end_strings: ["<|endoftext|>", "<extra_id_1>", "\x11"]
```

Comment:
IMO no, actually I think that ideally we would use
```
end_strings: ["<|endoftext|>"]
```
by default, because there is no reason to include in a generic open-source codebase some special tokens that are specific to NVIDIA's internal tokenizers and experiments settings. Especially since this slows down generation when those strings are not encoded as a single token (there was a customer where generation was >2x slower because of this).

The main problem is that unfortunately NeMo included `<extra_id_1>` by default in https://github.com/NVIDIA/NeMo/blob/59aa22d5f65009e8e5bac1d6e8d9bb5be1ec4162/nemo/collections/nlp/modules/common/text_generation_utils.py#L81 so we felt we had to do the same...

Reply:
Ah interesting, I didn't know this, okay, I will remove it